Userland/Services/SpiceAgent/VDAgent.cpp

```diff
@@ -0,0 +1,203 @@
+        auto data_buffer = ByteBuffer::create_uninitialized(message->size - sizeof(u32));
+
+        const auto total_bytes = message->size - sizeof(Clipboard);
+        auto bytes_copied = header.size - sizeof(Message) - sizeof(Clipboard);
+        memcpy(data_buffer.data(), clipboard_message->data, bytes_copied);
+
+        while (bytes_copied < total_bytes) {
+            ChunkHeader next_header;
+            read_n(&next_header, sizeof(ChunkHeader));
+            read_n(data_buffer.data() + bytes_copied, next_header.size);
+            bytes_copied += next_header.size;
+        }
+
+        m_just_set_clip = true;
+        auto anon_buffer = Core::AnonymousBuffer::create_with_size(data_buffer.size());
+        memcpy(anon_buffer.data<void>(), data_buffer.data(), data_buffer.size());
+        m_clipboard_connection.async_set_clipboard_data(anon_buffer, "text/plain", {});
+        break;
+    }
+    default:
+        dbgln("Unhandled message type {}", message->type);
+    }
+}
+
+void VDAgent::read_n(void* dest, size_t n)
+{
+    size_t bytes_read = 0;
+    while (bytes_read < n) {
+        int nread = read(m_fd, (u8*)dest + bytes_read, n - bytes_read);
+        if (nread > 0) {
+            bytes_read += nread;
+        } else if (errno == EAGAIN) {
+            continue;
+        } else {
+            dbgln("Failed to read: {}", errno);
+            return;
+        }
+    }
+}
+
+void VDAgent::send_message(ByteBuffer const& buffer)
+{
+    size_t bytes_written = 0;
+    while (bytes_written < buffer.size()) {
+        int result = write(m_fd, buffer.data() + bytes_written, buffer.size() - bytes_written);
+        if (result < 0) {
+            dbgln("Failed to write: {}", errno);
+            return;
+        }
+        bytes_written += result;
+    }
+}
+
+VDAgent::Message* VDAgent::initialize_headers(u8* data, size_t additional_data_size, MessageType type)
+{
+    new (data) ChunkHeader {
+        (u32)Port::Client,
+        (u32)(sizeof(Message) + additional_data_size)
+    };
+
+    auto* message = new (data + sizeof(ChunkHeader)) Message {
+        AGENT_PROTOCOL,
+        (u32)type,
+        0,
+        (u32)additional_data_size
+    };
+    return message;
+}
+
+ByteBuffer VDAgent::AnnounceCapabilities::make_buffer(bool request, const Vector<Capability>& capabilities)
+{
+    size_t required_size = sizeof(ChunkHeader) + sizeof(Message) + sizeof(AnnounceCapabilities);
+    auto buffer = ByteBuffer::create_uninitialized(required_size);
+    u8* data = buffer.data();
+
+    auto* message = initialize_headers(data, sizeof(AnnounceCapabilities), MessageType::AnnounceCapabilities);
+
+    auto* announce_message = new (message->data) AnnounceCapabilities {
+        request,
+        {}
+    };
+
+    for (auto& cap : capabilities) {
+        announce_message->caps[0] |= (1 << (u32)cap);
+    }
+
+    return buffer;
+}
+
+ByteBuffer VDAgent::ClipboardGrab::make_buffer(const Vector<ClipboardType>& types)
+{
+        VERIFY(types.size() > 0);
```

Comment:
indentation (instead of fixing this in the third commit please fix it in the original commit)